codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ShapeValueGenerator.java

```diff
@@ -0,0 +1,434 @@
+
+import java.util.logging.Logger;
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.node.ArrayNode;
+import software.amazon.smithy.model.node.BooleanNode;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.NodeVisitor;
+import software.amazon.smithy.model.node.NullNode;
+import software.amazon.smithy.model.node.NumberNode;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.StringNode;
+import software.amazon.smithy.model.shapes.CollectionShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeType;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.traits.EnumTrait;
+
+/**
+ * Generates a shape type declaration based on the parameters provided.
+ */
+public final class ShapeValueGenerator {
+    private static final Logger LOGGER = Logger.getLogger(ShapeValueGenerator.class.getName());
+
+    protected final Model model;
+    protected final SymbolProvider symbolProvider;
+
+    /**
+     * Initializes a shape value generator.
+     *
+     * @param model          the Smithy model references.
+     * @param symbolProvider the symbol provider.
+     */
+    public ShapeValueGenerator(Model model, SymbolProvider symbolProvider) {
+        this.model = model;
+        this.symbolProvider = symbolProvider;
+    }
+
+    /**
+     * Writes generation of a shape value type declaration for the given the parameters.
+     *
+     * @param writer writer to write generated code with.
+     * @param shape  the shape that will be declared.
+     * @param params parameters to fill the generated shape declaration.
+     */
+    public void writeShapeValueInline(GoWriter writer, Shape shape, Node params) {
+        switch (shape.getType()) {
+            case STRUCTURE:
+                structDeclShapeValue(writer, shape.asStructureShape().get(), () -> {
+                    params.accept(new ShapeValueNodeVisitor(writer, this, shape));
+                });
+                break;
+
+            case SET:
+            case LIST:
+                listDeclShapeValue(writer, (CollectionShape) shape, () -> {
+                    params.accept(new ShapeValueNodeVisitor(writer, this, shape));
+                });
+                break;
+
+            case MAP:
+                mapDeclShapeValue(writer, shape.asMapShape().get(), () -> {
+                    params.accept(new ShapeValueNodeVisitor(writer, this, shape));
+                });
+                break;
+
+            case UNION:
+            case DOCUMENT:
+                // TODO not supported yet
```

Comment:
Should this throw an exception for now?

Reply:
yeah can add throwing exception to make sure we see these later